src/omnipool/math.rs

```diff
@@ -0,0 +1,425 @@
+        stable_asset.0,
+        stable_asset.1
+    );
+
+    let delta_reserve_hp = current_reserve_hp
+        .checked_mul(delta_shares_hp)
+        .and_then(|v| v.checked_div(current_shares_hp))?;
+    let delta_hub_reserve_hp = delta_reserve_hp
+        .checked_mul(current_hub_reserve_hp)
+        .and_then(|v| v.checked_div(current_reserve_hp))?;
+
+    let delta_position_amount_hp = shares_removed_hp
+        .checked_mul(position_amount_hp)
+        .and_then(|v| v.checked_div(position_shares_hp))?;
+
+    let adjusted_asset_tvl_hp = if is_stable_asset {
+        // TODO: add test for this
+        stable_reserve_hp.checked_sub(delta_reserve_hp)?
+    } else {
+        stable_reserve_hp
+            .checked_mul(current_hub_reserve_hp.checked_sub(delta_hub_reserve_hp)?)
+            .and_then(|v| v.checked_div(stable_hub_reserve_hp))?
+    };
+
+    let adjusted_asset_tvl = to_balance!(adjusted_asset_tvl_hp).ok()?;
+    let delta_reserve = to_balance!(delta_reserve_hp).ok()?;
+    let delta_hub_reserve = to_balance!(delta_hub_reserve_hp).ok()?;
+    let delta_position_amount = to_balance!(delta_position_amount_hp).ok()?;
+
+    let delta_tvl = match adjusted_asset_tvl.cmp(&asset_state.tvl) {
+        Ordering::Greater => BalanceUpdate::Increase(adjusted_asset_tvl.checked_sub(asset_state.tvl)?),
+        Ordering::Less => BalanceUpdate::Decrease(asset_state.tvl.checked_sub(adjusted_asset_tvl)?),
+        Ordering::Equal => BalanceUpdate::Increase(Balance::zero()),
+    };
+
+    let hub_transferred = if current_price > position_price {
+        // LP receives some hub asset
+
+        // delta_q_a = -pi * ( 2pi / (pi + pa) * delta_s_a / Si * Ri + delta_r_a )
+        // note: delta_s_a is < 0
+
+        let price_sum = current_price.checked_add(&position_price)?;
+
+        let double_current_price = current_price.checked_mul(&FixedU128::from(2))?;
+
+        let p1 = double_current_price.checked_div(&price_sum)?;
+
+        let p2 = current_reserve_hp
+            .checked_mul(shares_removed_hp)
+            .and_then(|v| v.checked_div(current_shares_hp))?;
+
+        let p2 = to_balance!(p2).ok()?;
+
+        let p3 = p1.checked_mul_int(p2)?;
+
+        current_price.checked_mul_int(p3.checked_sub(delta_reserve)?)?
+    } else {
+        Balance::zero()
+    };
+
+    Some(LiquidityStateChange {
+        asset: AssetStateChange {
+            delta_reserve: Decrease(delta_reserve),
+            delta_hub_reserve: Decrease(delta_hub_reserve),
+            delta_shares: Decrease(delta_shares),
+            delta_protocol_shares: Increase(delta_b),
+            delta_tvl,
+        },
+        delta_imbalance: Increase(delta_reserve),
+        lp_hub_amount: hub_transferred,
+        delta_position_reserve: Decrease(delta_position_amount),
+        delta_position_shares: Decrease(shares_removed),
+    })
+}
+
+pub fn calculate_asset_tvl(asset_hub_reserve: Balance, stable_asset: (Balance, Balance)) -> Option<Balance> {
```

Comment:
I think that we could add tests for both this `calculate_asset_tvl` and  `calculate_delta_imbalance`, as they are parts of the public API.